src/transform/src/column_knowledge.rs

```diff
@@ -110,7 +111,96 @@ impl ColumnKnowledge {
+                        // TODO: This should respect the soft and hard max iteration
+                        // counts that will be enfoced when actually evaluating the
+                        // loop (see #18362 and #16800).
+                        if curr_iteration >= max_iterations {
+                            if curr_iteration >= 3 * let_rec_arity {
+                                soft_panic_or_log!(
+                                    "LetRec loop in ColumnKnowledge has not converged in 3 * |{}|",
+                                    let_rec_arity
+                                );
+                            }
+
+                            for (id, value) in zip_eq(ids.iter(), values.iter()) {
+                                let id = mz_expr::Id::Local(id.clone());
+                                let knowledge_new = vec![DatumKnowledge::top(); value.arity()];
+                                knowledge.insert(id, knowledge_new);
+                            }
+                            break;
+                        }
+
+                        // Check for condition (2).
+                        let mut change = false;
+                        for (id, mut value) in zip_eq(ids.iter(), values.iter().cloned()) {
+                            let id = mz_expr::Id::Local(id.clone());
+                            let value = &mut value;
+                            let next_knowledge = self.harvest(value, knowledge, knowledge_stack)?;
+                            let curr_knowledge = knowledge.get_mut(&id).unwrap();
+                            for (curr, next) in zip_eq(curr_knowledge.iter_mut(), next_knowledge) {
+                                let prev = curr.clone();
+                                curr.join_assign(&next);
+                                change |= prev != *curr;
+                            }
+                        }
+                        if !change {
+                            break;
+                        }
+
+                        curr_iteration += 1;
+                    }
+
+                    // Descend into the value and the body with the inferred knowledge.
+                    for value in values.iter_mut() {
+                        self.harvest(value, knowledge, knowledge_stack)?;
+                    }
+
+                    // Descend and return the knowledge from the body.
+                    let body_knowledge = self.harvest(body, knowledge, knowledge_stack)?;
+
+                    // Remove shadowed bindings. This is good hygiene, as
+                    // otherwise with nested LetRec blocks the `loop { ... }`
+                    // above will carry inner LetRec IDs across outer LetRec
+                    // iterations. As a consequence, the "no shadowing"
+                    // assertion at the beginning of this block will fail at the
+                    // inner LetRec for the second outer LetRec iteration.
+                    for id in ids.iter() {
```

Comment:
@ggevay / @mgree This was somewhat surprising to me, but at the end of the day a logical consequence of running a fixpoint loop. I believe that this "remove from context map" requirement will carry over to other transforms that have a loop in their `LetRec { .. }` case.